zk.js/src/utxo.ts

```diff
@@ -772,25 +816,100 @@ export class Utxo {
           account.encryptionKeypair.secretKey,
         );
 
+        // Return null if unable to decrypt
         if (!cleartext) {
-          return null;
+          return Result.Ok(null);
         }
+        // Convert decrypted cleartext to bytes
         const bytes = Buffer.from(cleartext);
-        return Utxo.fromBytes({
-          poseidon,
-          bytes,
-          account,
-          index,
-          appDataIdl,
-          assetLookupTable,
-          verifierProgramLookupTable,
-        });
+
+        // Return a decrypted UTXO
+        return Result.Ok(
+          Utxo.fromBytes({
+            poseidon,
+            bytes,
+            account,
+            index,
+            appDataIdl,
+            assetLookupTable,
+            verifierProgramLookupTable,
+          }),
+        );
       } else {
-        return null;
+        return Result.Ok(null);
       }
     }
   }
 
+  // TODO: unify compressed and includeAppData into a parsingConfig or just keep one
+  /**
+   * * @description Decrypts an utxo from an array of bytes by checking the UTXO prefix hash,
+   * * the first by 16 / 24 bytes of the commitment are the IV / nonce.
+   * @param {any} poseidon
+   * @param {Uint8Array} encBytes
+   * @param {Account} account
+   * @param {number} index
+   * @returns {Utxo | boolean }
+   */
+  static async decrypt({
+    poseidon,
+    encBytes,
+    account,
+    index,
+    merkleTreePdaPublicKey,
+    aes,
+    commitment,
+    appDataIdl,
+    compressed = true,
+    assetLookupTable,
+    verifierProgramLookupTable,
+  }: {
+    poseidon: any;
+    encBytes: Uint8Array;
+    account: Account;
+    index: number;
+    merkleTreePdaPublicKey?: PublicKey;
+    aes: boolean;
+    commitment: Uint8Array;
+    appDataIdl?: Idl;
+    compressed?: boolean;
+    assetLookupTable: string[];
+    verifierProgramLookupTable: string[];
+  }): Promise<Result<Utxo | null, UtxoError>> {
+    const prefixBytes = encBytes.slice(0, UTXO_PREFIX_LENGTH);
+    if (aes && this.checkPrefixHash({ account, commitment, prefixBytes })) {
+      const utxoResult = await Utxo.decryptUnchecked({
+        poseidon,
+        encBytes,
+        account,
+        index,
+        merkleTreePdaPublicKey,
+        aes,
+        commitment,
+        appDataIdl,
+        compressed,
+        assetLookupTable,
+        verifierProgramLookupTable,
+      });
+
+      // prefixHash matches but decryption fails so utxo is null -> Returns TRUE (COLLISION)
+      if (utxoResult.value) {
+        return utxoResult;
+      }
+
+      // prefixHash matches and decryption succeeds so utxo is good -> Returns UTXO (VALID)
```

Comment:
I will like this comment does not match the code any more, or did I misunderstand?